tests/frontier/opcodes/test_calldatacopy.py

```diff
@@ -0,0 +1,176 @@
+            ),
+            b"\x03",
+            Account(
+                storage={0x00: 0x00},
+            ),
+            Account(
+                storage={0x00: 0x00},
+            ),
+        ),
+        (
+            (
+                Op.CALLDATACOPY(
+                    dest_offset=0,
+                    offset=0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFA,
+                    size=0xFF,
+                )
+                + Op.SSTORE(key=0x0, value=Op.MLOAD(offset=0))
+                + Op.RETURN(offset=0, size=Op.MSIZE)
+            ),
+            b"\x04",
+            Account(storage={0x00: 0x00}),
+            Account(storage={0x00: 0x00}),
+        ),
+        (
+            (
+                Op.CALLDATACOPY(
+                    dest_offset=0,
+                    offset=0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFA,
+                    size=0x9,
+                )
+                + Op.SSTORE(key=0x0, value=Op.MLOAD(offset=0))
+                + Op.RETURN(offset=0, size=Op.MSIZE)
+            ),
+            b"\x05",
+            Account(storage={0x00: 0x00}),
+            Account(storage={0x00: 0x00}),
+        ),
+        (
+            (Op.SSTORE(key=0x1, value=0x1) + Op.PUSH1[0x1] + Op.PUSH1[0x2] + Op.CALLDATACOPY),
+            b"\x10",
+            Account(storage={0x01: 0x00}),
+            None,
+        ),
+        (
+            (
+                Op.JUMP(pc=0x5)
+                + Op.JUMPDEST
+                + Op.STOP
+                + Op.JUMPDEST
+                + Op.MSTORE8(offset=0x1F, value=0x42)
+                + Op.CALLDATACOPY(dest_offset=0x1F, offset=0x0, size=0x103)
+                + Op.MLOAD(offset=0x0)
+                + Op.DUP1
+                + Op.PUSH1[0x60]
+                + Op.JUMPI(pc=0x3, condition=Op.EQ)
+                + Op.SSTORE(key=0xFF, value=0xBADC0FFEE)
+            ),
+            b"\x11",
+            Account(storage={0xFF: 0xBADC0FFEE}),
+            None,
+        ),
+    ],
+    ids=[
+        "cdc 0 1 2",
+        "cdc 0 1 1",
+        "cdc 0 1 0",
+        "cdc 0 0 0",
+        "cdc 0 neg6 ff",
+        "cdc 0 neg6 9",
+        "underflow",
+        "sec",
+    ],
+)
+def test_calldatacopy(
+    state_test: StateTestFiller,
+    code: Bytecode,
+    fork: Fork,
+    tx_data: bytes,
+    pre: Alloc,
+    code_address_storage: Account,
+    to_address_storage: Account | None,
+):
+    """
+    Test `CALLDATACOPY` opcode.
+
+    Based on https://github.com/ethereum/tests/blob/ae4791077e8fcf716136e70fe8392f1a1f1495fb/src/GeneralStateTestsFiller/VMTests/vmTests/calldatacopyFiller.ym
+    """
+    code_address = pre.deploy_contract(code)
+    to = pre.deploy_contract(
+        code=(
+            Op.MSTORE(offset=0x0, value=0x1234567890ABCDEF01234567890ABCDEF0)
+            + Op.CALL(
+                0xFFFFFF, Op.ADD(0x1000, Op.CALLDATALOAD(offset=0x4)), 0x0, 0xF, 0x10, 0x20, 0x40
+            )
```

Comment:
The `Op.ADD(0x1000, Op.CALLDATALOAD(offset=0x4))` is unnecessary here.

The reason it was used in `ethereum/tests` it's because the parametrization happened using the transaction data, which `CALLDATALOAD` fetched and then it redirected the test flow to the correct contract.

But in EEST the return value of `pre.deploy_contract` is not entirely deterministic. It was working during fill because the default is indeed `0x1000`, and since the transaction had no data, `0x1000 + 0x00`, so we're good because we ended up in the correct contract. But in other contexts, such as live networks, the contract is completely different.

This can be changed to:
```suggestion
            + Op.CALL(
                gas=0xFFFFFF,
                address=code_address,
                value=0x0,
                args_offset=0xF,
                args_size=0x10,
                ret_offset=0x20,
                ret_size=0x40,
            )
```